src/util/http/HttpUtils.h

```diff
@@ -32,32 +32,42 @@ using tcp = boost::asio::ip::tcp;  // from <boost/asio/ip/tcp.hpp>
 namespace streams = ad_utility::streams;
 using ad_utility::httpUtils::httpStreams::streamable_body;
 
-/// The components of a URL. For example, the components of the URL
+/// Simple URL class that provides "getters" for the various URL components
+/// (some of the components are implicit, such as the port, so these are not
+/// really getters). For example, the components of the URL
 /// https://qlever.cs.uni-freiburg.de/api/wikidata are:
 ///
 /// protocol: HTTPS
 /// host:     qlever.cs.uni-freiburg.de
 /// port:     443 (implicit)
 /// target:   /api/wikidata .
 ///
-/// NOTE: `host` and `target` could be `std::string_view` because they are parts
-/// of the given URL. However, `port` can be implicit, so we need a
-/// `std::string` here (and it's not an `int` because the Beast functions ask
-/// for the port as a string). Since URLs are short and we do not handle large
-/// numbers of URLs, the overhead of the string copies are negligible.
-struct UrlComponents {
+class Url {
+ public:
+  enum class Protocol { HTTP, HTTPS };
+
+ private:
+  Protocol protocol_;
+  std::string host_;
+  std::string port_;
+  std::string target_;
+
+ public:
   // Construct from given URL.
-  UrlComponents(const std::string_view url);
-  // Members.
-  enum Protocol { HTTP, HTTPS } protocol;
-  std::string host;
-  std::string port;
-  std::string target;
-  // For testing.
-  friend std::ostream& operator<<(std::ostream& os, const UrlComponents& uc) {
-    return os << "UrlComponents("
-              << (uc.protocol == Protocol::HTTP ? "http" : "https") << ", "
-              << uc.host << ", " << uc.port << ", " << uc.target << ")";
+  Url(const std::string_view url);
+  // The protocol: one of Protocol::HTTP or Protocol::HTTPS.
+  Protocol protocol() const { return protocol_; }
+  // The host; this is always a substring of the given URL.
+  std::string host() const { return host_; }
+  // The port; inferred from the protocol if not specified explicitly (80 for
+  // HTTP, 443 for HTTPS).
+  std::string port() const { return port_; }
+  // The target; this is a substring of the given URL, except when it's empty in
+  // the URL, then it's "/".
+  std::string target() const { return target_; }
```

Comment:
Should return `const string&` (I think you often only need to read them, and now we ALWAYS have a copy).